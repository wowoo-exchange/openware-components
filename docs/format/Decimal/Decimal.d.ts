import * as React from 'react';
interface DecimalProps {
    /**
     * Number of digits after dot
     */
    fixed: number;
    /**
     * Number to format
     */
    children?: string | number;
    /**
     * Number of highlighted digits (without opacity) starting from the end
     * By default only integer part of the number is highlighted
     */
    highlightedNumbersAmount?: number;
}
declare class Decimal extends React.Component<DecimalProps> {
    static format(value: DecimalProps['children'], fixed: number): string;
    static getNumberBeforeDot(value: DecimalProps['children'], fixed: number): number;
    static getNumberAfterDot(value: DecimalProps['children'], fixed: number): string | undefined;
    render(): JSX.Element;
    private highlightNumbers;
}
export { Decimal, DecimalProps, };
