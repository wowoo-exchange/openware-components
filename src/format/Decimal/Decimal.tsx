import * as React from 'react';

interface DecimalProps {
    /**
     * Number of digits after dot
     */
    fixed: number;
    /**
     * Number to format
     */
    children?: string | number;
    /**
     * Number of highlighted digits (without opacity) starting from the end
     * By default only integer part of the number is highlighted
     */
    highlightedNumbersAmount?: number;
}

class Decimal extends React.Component<DecimalProps> {
    public static format(value: DecimalProps['children'], fixed: number) {
        const baseNumber = 10;
        let result = (typeof value === 'undefined') ? '0' :
            (Math.floor(+value * Math.pow(baseNumber, fixed)) / Math.pow(baseNumber, fixed)).toString();
        if (result.indexOf('.') === -1) {
            result += '.';
        }

        while (result.slice(result.indexOf('.')).length <= fixed) {
            result += '0';
        }

        return result;
    }

    public static getNumberBeforeDot(value: DecimalProps['children'], fixed: number) {
        const num = Decimal.format(value, fixed);

        return Math.floor(+num);
    }

    public static getNumberAfterDot(value: DecimalProps['children'], fixed: number) {
        if (fixed === 0) {
            return;
        }

        const str = Decimal.format(value, fixed);

        return str.slice(str.indexOf('.'));
    }

    public render() {
        const { children, fixed } = this.props;
        if (this.props.highlightedNumbersAmount) {
            return (this.highlightNumbers(Decimal.format(children, fixed), this.props.highlightedNumbersAmount));
        } else {
            return (
                <React.Fragment>
                    <span>{Decimal.getNumberBeforeDot(children, fixed)}</span>
                    <span className="cr-decimal__opacity">{Decimal.getNumberAfterDot(children, fixed)}</span>
                </React.Fragment>
            );
        }
    }

    private highlightNumbers = (value: string, amount: number) => {
        let highlighted = '';
        let i = value.length - 1;
        let highlightedAmount = value.length - amount;

        for (i; i >= highlightedAmount; i--) {
            if (value[i] === '.') {
                highlighted = `.${highlighted}`;
                i--;
                highlightedAmount--;
            }

            highlighted = value[i] + highlighted;
        }

        return (
            <React.Fragment>
                <span className="cr-decimal__opacity">{value.slice(0, value.length - highlighted.length)}</span>
                <span>{highlighted}</span>
            </React.Fragment>
        );
    };
}

export {
    Decimal,
    DecimalProps,
};
