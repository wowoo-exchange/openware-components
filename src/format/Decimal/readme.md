Decimal format example:

```js
const fixed = 2;

<Decimal fixed={fixed} highlightedNumbersAmount={3}>1234.56789</Decimal>
```