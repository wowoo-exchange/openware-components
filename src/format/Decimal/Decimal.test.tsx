import { shallow } from 'enzyme';
import * as React from 'react';
import { Decimal, DecimalProps } from './Decimal';

const defaultProps: DecimalProps = {
    fixed: 8,
};

// tslint:disable no-magic-numbers
const setup = (props: Partial<DecimalProps> = {}) =>
    shallow(<Decimal {...{ ...defaultProps, ...props }}>123.32030200230000</Decimal>);

describe('Decimal', () => {
    it('should render', () => {
        const wrapper = setup();
        expect(wrapper).toMatchSnapshot();
    });

    it('should have className for number after dot', () => {
        const wrapper = setup();
        expect(wrapper.find('span').last().hasClass('cr-decimal__opacity')).toBeTruthy();
    });

    it('should handle string child', () => {
        const wrapper = setup({ children: '123.3203020023' });
        expect(wrapper).toMatchSnapshot();
    });

    it('should handle number child', () => {
        const wrapper = setup({ children: 123.3203020023 });
        expect(wrapper).toMatchSnapshot();
    });

    it('should handle undefined value', () => {
        const wrapper = shallow(<Decimal fixed={8}>{undefined}</Decimal>);
        expect(wrapper.find('span').first().props().children).toEqual(0);
        expect(Decimal.format(undefined, 3)).toEqual('0.000');
    });

    it('should handle format with function Decimal.format', () => {
        const formatWithFunc = Decimal.format(123.3203020023, 3);
        expect(formatWithFunc).toEqual('123.320');
    });

    it('should handle format with function Decimal.getNumberBeforeDot', () => {
        const formatWithFunc = Decimal.getNumberBeforeDot(123.3203020023, 3);
        expect(formatWithFunc).toEqual(123);
    });

    it('should handle format with function Decimal.getNumberAfterDot', () => {
        const formatWithFunc = Decimal.getNumberAfterDot(123.3203020023, 3);
        expect(formatWithFunc).toEqual('.320');
    });

    it('should handle zero fixed value', () => {
        const wrapper = setup({ fixed: 0 });
        expect(wrapper.find('span').first().props().children).toEqual(123);
    });

    it('should handle highlightedNumbersAmount prop', () => {
        const wrapper = setup({ fixed: 2, highlightedNumbersAmount: 3 });
        expect(wrapper.find('span').first().hasClass('cr-decimal__opacity')).toBeTruthy();
        expect(wrapper.find('span').first().props().children).toEqual('12');
        expect(wrapper.find('span').last().props().children).toEqual('3.32');
    });
});
