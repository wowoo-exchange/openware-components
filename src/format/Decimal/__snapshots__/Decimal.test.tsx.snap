// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Decimal should handle number child 1`] = `
<Fragment>
  <span>
    123
  </span>
  <span
    className="cr-decimal__opacity"
  >
    .32030200
  </span>
</Fragment>
`;

exports[`Decimal should handle string child 1`] = `
<Fragment>
  <span>
    123
  </span>
  <span
    className="cr-decimal__opacity"
  >
    .32030200
  </span>
</Fragment>
`;

exports[`Decimal should render 1`] = `
<Fragment>
  <span>
    123
  </span>
  <span
    className="cr-decimal__opacity"
  >
    .32030200
  </span>
</Fragment>
`;
